lib/ui/window.dart

```diff
@@ -143,14 +161,234 @@ class Locale {
-  /// optional.
-  ///
-  /// The values are _case sensitive_, and should match the case of the relevant
-  /// subtags in the [IANA Language Subtag
-  /// Registry](https://www.iana.org/assignments/language-subtag-registry/language-subtag-registry).
-  /// Typically this means the primary language subtag should be lowercase and
-  /// the region subtag should be uppercase.
-  const Locale(this._languageCode, [ this._countryCode ]) : assert(_languageCode != null);
+  /// optional. The values are _case sensitive_, and must match the values
+  /// listed in
+  /// http://unicode.org/repos/cldr/tags/latest/common/validity/language.xml and
+  /// http://unicode.org/repos/cldr/tags/latest/common/validity/region.xml.
+  ///
+  /// This method only produces standards-compliant instances if valid language
+  /// and country codes are provided. Deprecated subtags will be replaced, but
+  /// incorrectly cased strings are not corrected.
+  const Locale(
+    this._languageCode, [
+    this._countryCode,
+  ]) : assert(_languageCode != null),
+       this._scriptCode = null,
+       this._variants = null,
+       this._extensions = null;
+
+  /// Creates a new Locale object with the specified parts.
+  ///
+  /// This is for internal use only. All fields must already be normalized, must
+  /// already be canonicalized. This method does not modify parameters in any
+  /// way or do any syntax checking.
+  ///
+  /// * language, script and region must be in canonical form.
+  /// * iterating over variants must provide variants in alphabetical order.
+  /// * The extensions map must contain only valid key/value pairs. "u" and "t"
+  ///   keys must be present, with an empty string as value, if there are any
+  ///   subtags for those singletons.
+  Locale._internal(
+    String language, {
+    String script,
+    String region,
+    collection.SplayTreeSet<String> variants,
+    collection.SplayTreeMap<String, String> extensions,
+  }) : assert(language != null),
+       assert(language.length >= 2),
+       assert(language.length <= 8),
+       assert(language.length != 4),
+       assert((script ?? 'Xxxx').length == 4),
+       assert((region ?? 'XX').length >= 2),
+       assert((region ?? 'XX').length <= 3),
+       _languageCode = language,
+       _scriptCode = script,
+       _countryCode = region,
+       _variants = collection.LinkedHashSet<String>.from(variants ?? <String>[]),
+       _extensions = collection.LinkedHashMap<String, String>.from(extensions ?? <String, String>{});
+
+  /// Parses [Unicode Locale
+  /// Identifiers](https://www.unicode.org/reports/tr35/#Identifiers).
+  ///
+  /// This method does not parse all BCP 47 tags. See [BCP 47
+  /// Conformance](https://www.unicode.org/reports/tr35/#BCP_47_Conformance) for
+  /// details.
+  ///
+  /// TODO: we should try to support parsing all BCP 47 tags, even if we produce
+  /// only Unicode BCP47 Locale Identifiers as output.
+  factory Locale.parse(String localeId) {
```

Comment:
nit: convention is to make this a static method, not a constructor. (the only real effect is where the member ends up in the documentation)

Reply:
Conventionally, when would you use a factory constructor? (By Flutter convention, never? I suspect this isn't Dart convention at this time?)
"factory" is not mentioned in https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo

Reply:
Considering the constructors of List for example:
https://api.dartlang.org/stable/2.0.0/dart-core/List-class.html#constructors

Locale.parse could be thought of as a core constructor, something I would have proposed as the unnamed constructor, if it didn't already exist. (Hence leaving this discussion point open.)

Reply:
You typically use a constructor when it isn't supposed to ever fail (unless misused).

But `parse` is a particularly special case because there's dozens of classes in Dart that all have a `parse` (and in some cases `tryParse`) method, and by and large they all have the same signature and are all static methods.

Reply:
Also looks like a thing that could be changed and changed back easily in the future without breaking anything (as you mention, it being essentially a documentation change). Changing it to a static method. Thanks!